plugins/cloud-gce/src/main/java/org/elasticsearch/cloud/gce/network/GceNameResolver.java

```diff
@@ -0,0 +1,114 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cloud.gce.network;
+
+import org.elasticsearch.cloud.gce.GceComputeService;
+import org.elasticsearch.cloud.gce.GceComputeService.Fields;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.network.NetworkService.CustomNameResolver;
+import org.elasticsearch.common.settings.Settings;
+
+import java.io.IOException;
+import java.net.InetAddress;
+
+/**
+ * <p>Resolves certain GCE related 'meta' hostnames into an actual hostname
+ * obtained from gce meta-data.</p>
+ * Valid config values for {@link GceAddressResolverType}s are -
+ * <ul>
+ * <li>_gce_ - maps to privateIp</li>
+ * <li>_gce:privateIp_</li>
+ * <li>_gce:hostname_</li>
+ * </ul>
+ */
+public class GceNameResolver extends AbstractComponent implements CustomNameResolver {
+
+    private final GceComputeService gceComputeService;
+    private final Integer networkCard;
+
+    /**
+     * enum that can be added to over time with more meta-data types
+     */
+    private enum GceAddressResolverType {
+
+        // some less verbose defaults
+        PRIVATE_DNS("gce:hostname", "hostname"),
+        PRIVATE_IP("gce:privateIp", "network-interfaces/{{network}}/ip"),
+        GCE("gce", PRIVATE_IP.gceName);
+
+        final String configName;
+        final String gceName;
+
+        GceAddressResolverType(String configName, String gceName) {
+            this.configName = configName;
+            this.gceName = gceName;
+        }
+    }
+
+    /**
+     * Construct a {@link CustomNameResolver}.
+     */
+    public GceNameResolver(Settings settings, GceComputeService gceComputeService) {
+        super(settings);
+        this.gceComputeService = gceComputeService;
+        // We by default extract metadata from network card 0
+        // network-interfaces/0/ip
+        this.networkCard = settings.getAsInt(Fields.NETWORK_CARD, 0);
+    }
+
+    /**
+     * @param type the gce hostname type to discover.
+     * @param networkCard
+     * @return the appropriate host resolved from gce meta-data.
+     * @see CustomNameResolver#resolveIfPossible(String)
+     */
+    private InetAddress[] resolve(GceAddressResolverType type, Integer networkCard) {
+        try {
+            // We replace network placeholder with networkCard value
+            String gceMetadataPath = Strings.replace(type.gceName, "{{network}}", networkCard.toString());
+            String metadataResult = gceComputeService.networkAddress(gceMetadataPath);
+            if (metadataResult == null || metadataResult.length() == 0) {
+                logger.error("no gce metadata returned for {}", type.configName);
+                return null;
```

Comment:
@bleskes Actually we can't throw an exception here because we will never be able to use for example `local`. 

Or we need to write differently that piece of code: https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/network/NetworkService.java#L169-L174